routers/common/logger_v2.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package common
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	gitea_context "code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// NewLoggerHandlerV2 is a handler that will log the routing to the default gitea log
+// About performance:
+// In v1, every request outputs 2 logs (Started/Completed)
+// In v2 (this), every request only outputs one log,
+//    all runtime reflections of handler functions are cached
+//	  the mutexes work in fast path for most cases (atomic incr) because there is seldom concurrency writings.
+// So generally speaking, the `logger context` doesn't cost much, using v2 with `logger context` will not affect performance
+// Instead, the performance may be improved because now only 1 log is outputted for each request.
+func NewLoggerHandlerV2() func(next http.Handler) http.Handler {
+	lh := logContextHandler{
+		requestRecordMap: map[uint64]*logRequestRecord{},
+	}
+	lh.startSlowQueryDetector(3 * time.Second)
+
+	lh.printLog = func(trigger LogRequestTrigger, reqRec *logRequestRecord) {
+		if trigger == LogRequestStart && !log.DEBUG.IsEnabledOn(setting.RouterLogLevel) {
+			// for performance, if the START message shouldn't be logged, we just return as early as possible
+			// developers could set both `log.LEVEL=debug` and `log.ROUTER_LOG_LEVEL=debug` to get the "started" request messages.
```

Comment:
This changes the semantics of ROUTER_LOG_LEVEL and the cheatsheet/app.example.ini needs to be updated

Reply:
I will revert the default behavior.